protocol/contracts/core/MultiOtcOffer.sol

```diff
@@ -0,0 +1,187 @@
+    }
+
+    /// @notice Owner can pause user swaps from occuring
+    function pause() external override onlyElevatedAccess {
+        _pause();
+    }
+
+    /// @notice Owner can unpause so user swaps can occur
+    function unpause() external override onlyElevatedAccess {
+        _unpause();
+    }
+
+    /// @notice Swap `userSellToken` for `userBuyToken`, at the `offerPrice`
+    function swap(bytes32 marketId, uint256 sellTokenAmount) external override whenNotPaused returns (uint256 buyTokenAmount) {
+        if (!_otcMarketIds.contains(marketId)) { revert InvalidMarketId(marketId); }
+        if (sellTokenAmount == 0) revert CommonEventsAndErrors.ExpectedNonZero();
+        OTCMarketInfo memory marketInfo = otcMarketInfo[marketId];
+        buyTokenAmount = quote(marketId, sellTokenAmount);
+        address _fundsOwner = marketInfo.fundsOwner;
+        IERC20Metadata _userSellToken = marketInfo.userSellToken;
+        IERC20Metadata _userBuyToken = marketInfo.userBuyToken;
+        emit Swap(msg.sender, _fundsOwner, sellTokenAmount, buyTokenAmount);
+
+        _userSellToken.safeTransferFrom(msg.sender, _fundsOwner, sellTokenAmount);
+        _userBuyToken.safeTransferFrom(_fundsOwner, msg.sender, buyTokenAmount);
+    }
+
+    /// @notice How many `userBuyToken` you would receive given an amount of `sellTokenAmount`
+    function quote(bytes32 marketId, uint256 sellTokenAmount) public override view returns (uint256 buyTokenAmount) {
+        if (!_otcMarketIds.contains(marketId)) { revert InvalidMarketId(marketId); }
+        OTCMarketPriceParams memory marketPriceParam = otcMarketPriceParams[marketId];
+        buyTokenAmount = marketPriceParam.offerPricingToken == OfferPricingToken.UserBuyToken
+            ? sellTokenAmount * marketPriceParam.offerPrice / marketPriceParam.scalar
+            : sellTokenAmount * marketPriceParam.scalar / marketPriceParam.offerPrice;
+    }
+
+     /**
+     * @notice The available funds for a user swap is goverend by the amount of `userBuyToken` that
+     * the `fundsOwner` has available.
+     * @dev The minimum of the `fundsOwner` balance of `userBuyToken`, and the spending 
+     * approval from `fundsOwner` to this OtcOffer contract.
+     */
+    function userBuyTokenAvailable(bytes32 _marketId) external override view returns (uint256) {
+        if (!_otcMarketIds.contains(_marketId)) { revert InvalidMarketId(_marketId); }
+        OTCMarketInfo memory marketInfo = otcMarketInfo[_marketId];
+        address _fundsOwner = marketInfo.fundsOwner;
+        uint256 _balance = marketInfo.userBuyToken.balanceOf(_fundsOwner);
+        uint256 _allowance = marketInfo.userBuyToken.allowance(_fundsOwner, address(this));
+        return _balance < _allowance
+            ? _balance
+            : _allowance;
+    }
+
+    function getOtcMarketIds() external view override returns (bytes32[] memory) {
+        return _otcMarketIds.values();
```

Comment:
If we had a marketId and wanted to know the tokens, is that possible? Maybe store a reverse mapping? Or is the idea that you can only know from looking at past Events?

Reply:
okay, will update